src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3800,46 +3769,66 @@ public Object stringCharacterIndex(DynamicObject string, DynamicObject pattern,
         }
     }
 
-    @Primitive(name = "string_byte_index_from_char_index", needsSelf = false, lowerFixnum = 2)
-    @ImportStatic(StringGuards.class)
-    public static abstract class StringByteIndexFromCharIndexNode extends PrimitiveArrayArgumentsNode {
+    @NodeChildren({
+            @NodeChild("rope"),
+            @NodeChild("startByteOffset"),
+            @NodeChild("characterIndex")
+    })
+    @ImportStatic({ RopeGuards.class, StringGuards.class, StringOperations.class })
+    public static abstract class ByteIndexFromCharIndexNode extends RubyNode {
 
-        public abstract Object executeFindByteIndex(DynamicObject string, int characterIndex);
+        public static ByteIndexFromCharIndexNode create() {
+            return StringNodesFactory.ByteIndexFromCharIndexNodeGen.create(null, null, null);
+        }
 
-        @Specialization(guards = "characterIndex < 0")
-        protected Object byteIndexNegativeIndex(DynamicObject string, int characterIndex) {
-            throw new RaiseException(
-                    getContext().getCoreExceptions().argumentError(
-                            coreStrings().CHARACTER_INDEX_NEGATIVE.getRope(), this));
+        public abstract int execute(Rope rope, int startByteOffset, int characterIndex);
+
+        @Specialization(guards = { "characterIndexInBounds(rope, characterIndex)", "rope.isSingleByteOptimizable()" })
+        protected int singleByteOptimizable(Rope rope, int startByteOffset, int characterIndex,
+                @Cached("createIdentityProfile()") IntValueProfile offsetProfile) {
+            return offsetProfile.profile(startByteOffset) + characterIndex;
```

Comment:
How does this work?
Is `startByteOffset` jut always added? Then the caller should probably do it.
Or is "count the number of bytes for `characterIndex` characters after `startByteOffset`"?
Maybe `nCharacters` would be clearer then.

Reply:
In this case the caller could add it. But in the `multiBytes` case it affects the starting point for the character search and because that's for variable-width MBC encodings, the caller can't do anything so straightforward as that.